cli/src/commands/rebase.rs

```diff
@@ -251,9 +268,17 @@ fn rebase_descendants(
     // `rebase_descendants` takes care of sorting in reverse topological order, so
     // no need to do it here.
     for old_commit in old_commits {
-        rebase_commit(settings, tx.mut_repo(), old_commit, new_parents)?;
+        rebase_commit_with_options(
+            settings,
+            tx.mut_repo(),
+            old_commit,
+            new_parents,
+            &rebase_options,
+        )?;
     }
-    let num_rebased = old_commits.len() + tx.mut_repo().rebase_descendants(settings)?;
+    let num_rebased = old_commits.len()
+        + tx.mut_repo()
+            .rebase_descendants_with_options(settings, rebase_options)?;
     writeln!(ui.stderr(), "Rebased {num_rebased} commits")?;
```

Comment:
It would be nice if the skipped commits are listed (or the number of abandoned commits is reported.)

For the `--skip-empty` (not `emptied`) behavior, maybe we can abandon empty commits as a separate pass. They can also be tracked by the rebaser (or abuse the record stored in mut_repo.) I don't know which is better overall, though.

Reply:
I agree about reporting skipped commits.

I also considered abandoning empty commits as a separate pass and I'm also not sure which is better :) One advantage is that it seems cleaner by keeping that logic separate. One disadvantage is that it's slower (probably fine) and wastes some disk space by creating two commits for each descendant. Since it's already implemented inside `DescendantRebaser` and it's not complicating it terribly much, I'd say we should just keep it that way for now at least.

Reply:
I think that if we report skipped commits, we should do so in a seperate PR. It'll require a nontrivial amount of plumbing changes to ensure that things that return a usize for the number of commits modified instead return a struct.

Reply:
> I think that if we report skipped commits, we should do so in a seperate PR.

Agreed. This isn't a blocker. I was just wondering what would be needed to report such stats.

Reply:
Currently, after `rebase_descendants` is run, the only thing saved from the run of DescendantRebase is the list of rebased commits. It would be natural to preserve the list of abandoned commits as well.

-------------------

Here are some additional, more vague, thoughts.

I think I'll soon want to preserve all the information, especially the `new_parents` part of the DescendantRebaser. I'm not sure whether the best way to encapsulate all that information is the way DescendantRebaser keeps track of it now.

This seems like it'll make `jj rebase -r --after` easier. I'd want to add a function to rebase a single commit based on substitution rules that come from previous run of DescendatRebaser. Ideally, I'd want to make DescendatRebaser itself, `jj new --after/--before`, and `jj rebase -r --after/--before` to use such a function.

It's also possible that this is unnecessary and we should just let this functionality be implemented separately in 2-3 places.

Reply:
> I think I'll soon want to preserve all the information

Would this help for `undo`?

Reply:
> Would this help for undo?

I don't think so, at least not the way I was thinking of it.

It's a curious idea. Storing such information in the operation log in some way seems like it could be helpful and powerful. It seems like a large project even to figure out exactly how we'd use it or present such information to the user. This seems related to the a short and inconclusive [Discord discussion on this topic](https://discord.com/channels/968932220549103686/968932220549103689/1177006969597481084) I just had with Martin.

Reply:
I'll mark this as resolved, since it's not required for this PR